src/oversampling_methods/borderline_smote1/borderline_smote1.jl

```diff
@@ -0,0 +1,239 @@
+               k = 5, ratios = Dict(0=>1.0, 1=> 0.9, 2=>0.8), rng = 42)
+
+julia> Imbalance.checkbalance(y)
+2: ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 38 (79.2%) 
+1: ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 43 (89.6%) 
+0: ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 48 (100.0%) 
+```
+
+# MLJ Model Interface
+
+Simply pass the keyword arguments while initiating the `BorderlineSMOTE1` model and pass the 
+    positional arguments to the `transform` method. 
+
+```julia
+using MLJ
+BorderlineSMOTE1 = @load BorderlineSMOTE1 pkg=Imbalance
+
+# Wrap the model in a machine
+oversampler = BorderlineSMOTE1(m=3, k=5, ratios=Dict(0=>1.0, 1=> 0.9, 2=>0.8), rng=42)
+mach = machine(oversampler)
+
+# Provide the data to transform (there is nothing to fit)
+Xover, yover = transform(mach, X, y)
+```
+You can read more about this `MLJ` interface [here]().
+
+
+
+# TableTransforms Interface
+
+This interface assumes that the input is one table `Xy` and that `y` is one of the columns. Hence, an integer `y_ind`
+    must be specified to the constructor to specify which column `y` is followed by other keyword arguments. 
+    Only `Xy` is provided while applying the transform.
+
+```julia
+using Imbalance
+using Imbalance.TableTransforms
+
+# Generate imbalanced data
+num_rows = 200
+num_features = 5
+y_ind = 3
+Xy, _ = generate_imbalanced_data(num_rows, num_features; 
+                                 probs=[0.5, 0.2, 0.3], insert_y=y_ind, rng=42)
+
+# Initiate BorderlineSMOTE1 Oversampler model
+oversampler = BorderlineSMOTE1(y_ind; m=3, k=5, 
+              ratios=Dict(0=>1.0, 1=> 0.9, 2=>0.8), rng=42)
+Xyover = Xy |> oversampler                              
+Xyover, cache = TableTransforms.apply(oversampler, Xy)    # equivalently
+```
+The `reapply(oversampler, Xy, cache)` method from `TableTransforms` simply falls back to `apply(oversample, Xy)` and the `revert(oversampler, Xy, cache)`
+reverts the transform by removing the oversampled observations from the table.
+
+
+# References
+[1] Han, H., Wang, W.-Y., & Mao, B.-H. (2005). Borderline-SMOTE: A new over-sampling method in imbalanced data sets learning. 
+    In D.S. Huang, X.-P. Zhang, & G.-B. Huang (Eds.), Advances in Intelligent Computing (pp. 878-887). Springer. 
+"""
+function borderline_smote1(
+	X::AbstractMatrix{<:AbstractFloat},
+	y::AbstractVector;
+    m::Integer = 5,
+	k::Integer = 5,
+	ratios = 1.0,
+	rng::Union{AbstractRNG, Integer} = default_rng(),
+    try_perserve_type::Bool = true,
+)
+    # this function adjust generic_oversampling to use in borderline smote
+    rng = rng_handler(rng)
+    X = transpose(X)
+    filter = borderline1_filter(X, y; m)
+    X1, y1 = X[:, filter], y[filter]       # filtered points should be oversampled
+
+    # give the user an idea about the filtered data.
+    y1_stats = match(r"\((.*?)\)", string(countmap(y1))).captures[1]
+    @info "After filtering, the mapping from each class to number of borderline points is ($y1_stats)."
```

Comment:
Minor commentary: I have found that users often want the option of tunring this sort of logging off. In a larger pipeline, this can contribute to a lot distracting noise. In MLJ models, we generally do something like

```julia
verbosity > 0 && @info ...
``` 
but that's not helpful here because your core function does not have `verbosity` as an arg or kwarg. Perhaps you could make verbosity a `kwarg` here and have your MLJ interface pass `verbosity` on. 

(In the best scenario, this kind of information goes into a model's report. But I guess that will require substantial refactoring of all your code for consistency.)

Fine if you disagree or if you want to kick the can on this one and open an issue instead.

Reply:
> but that's not helpful here because your core function does not have verbosity as an arg or kwarg. Perhaps you could make verbosity a kwarg here and have your MLJ interface pass verbosity on.

I find this satisfactory to me.